### PR TITLE
An easy way to simplify the channel methods

### PR DESCRIPTION
**Pull request by jcready**
Please read the [comments in the code](/NGPVAN/amqp.channel/blob/558a20ddb8723cce63fa4529e21cb3e9eee00415/simplify.js#L13-L40) ~~or [see other discussion](/NGPVAN/amqp.node.utils/pull/1#issuecomment-77698030) for explanation~~.

The `channel` object resolved by the returned `Promise` will behave differently from a normal `channel` object returned by the amqplib library in a few (hopefully convenient) ways:

1. The `consume`, `publish`, and `sendToQueue` channel methods have been changed to  explicitly handle JSON.
2. The `publish` and `sendToQueue` methods have been "promisified" in a way that will still provide information to know whether or not the write buffer is full (and therefore, whether or not you should continue writing to it) by adding an additional `ok` boolean property to the promise.
3. A `channel` consumer callback will no longer receive `null` when that consumer had been cancelled by Rabbit MQ. Instead, the `channel` object will emit a `'cancelled'` event with all the arguments passed to the `channel.consume()` call for the consumer that was cancelled.

#### Usage:

Automatic translation of JS object to JSON string to Buffer for sending/publishing:

```javascript
channel.sendToQueue('someQueue', { hello: 'world' });
channel.publish('someExchange', 'routingKey', { hello: 'world' });
```

Promisification of `sendToQueue` and `publish` methods:

```javascript
return channel.sendToQueue('someQueue', { hello: 'world' }).then(function(){
  return channel.publish('someExchange', 'routingKey', { hello: 'world' });
});
```

Automatic translation of message Buffer to JSON string to JS object for consuming:

```javascript
channel.sendToQueue('someQueue', { hello: 'world' });
channel.consume('someQueue', function(parsedMessage, originalMessage){
  console.log('hello', parsedMessage.hello); // => hello world
  channel.ack(originalMessage);
});
```

Handling a consumer getting cancelled by Rabbit MQ:

```javascript
channel.on('cancelled', function(queue, callback, options){
  // When the consumer below gets cancelled by Rabbit MQ
  console.log(queue, callback.name, options); // 'someQueue', 'onMessage', { noAck: true }
});
channel.consume('someQueue', function onMessage(parsedMessage, originalMessage){
  console.log(parsedMessage);
}, { noAck: true });
```

The `ok` property on the promises returned by the `sendToQueue` and `publish` methods:

```javascript
var sent = channel.sendToQueue('someQueue', { hello: 'world' });
if (sent.ok) {
  // continue sending
} else {
  // maybe pause sending until unblocked?
  channel.once('drain', function(){
    // continue sending
  });
}
```

### CODE REVIEW

**Comment by jcready** [comment]
@NGPVAN/violet @NGPVAN/freedom

File: channel.js [hunk 1]
```diff
@@ -1,8 +1,9 @@
 var noop = function(){},
     amqp = require('amqplib'),
-    Promise = require('bluebird');
+    Promise = require('bluebird'),
+    simplify = require('./simplify');
 
-module.exports = function createChannel(url, assertions, log){
+module.exports = function createChannel(url, assertions, log, makeSimple){
```
Comment (jcready) [inline channel.js]: I'm considering making this a major release and having this module perform the simplification by default.

**Comment by ojintoad** [comment]
if hyperlinks on this PR don't work it's becuase a repo that was originally going to house this work was deleted.

File: channel.js (hunk 1 above)
Comment (ojintoad) [inline channel.js]: go for it, i'll write up tech debt to upgrade our modules to use the upgraded package.

**Comment by bwestergard** [comment]
Can we put these docs in the README?

**Comment by bwestergard** [comment]
Other than that, looks good.

**Comment by jcready** [comment]
Yes, I will add some better documentation.

**Comment by ojintoad** [comment]
If checking promise.ok and waiting until unblocked is going to be a consistent pattern then that should also be wrapped up?

**Comment by jcready** [comment]
@ojintoad it isn't a requirement to use that pattern, but it is a pattern you'd probably want to use if you were publishing at a high frequency. From the Node's documentation on the return value of a `Writable.write()` call:

> The return value indicates if you should continue writing right now. If the data had to be buffered internally, then it will return `false`. Otherwise, it will return `true`.
>
> **This return value is strictly advisory.** You MAY continue to write, even if it returns `false`. However, writes will be buffered in memory, so it is best not to do this excessively. Instead, wait for the `drain` event before writing more data.

**Comment by ojintoad** [comment]
So basically we should start using it when we start noticing problems and then we can write a wrapper for it.

**Comment by jcready** [comment]
Yes, I had to write a weird wrapper for it when trying to "import" the initial 19 million rows from the CSV. I was streaming the CSV (line-by-line), and was publishing out messages to found.ly or not.found.ly depending on the first of second column of the row. Streaming the CSV was **MUCH** faster than publishing messages over the AMQP channel, so I had to setup a custom `publish` method that would basically "pause" the CSV streaming until either the call to `channel.publish()` returned true or (if it returned false) wait for the channel's `"drain"` event to fire:

```javascript
var message = { source: 'initial.ly' }, msg, row;
var reading = new Reader('../../Downloads/full.contact.csv');
reading.on('line', readLn);
reading.on('end', success);
reading.on('error', error);
channel.on('error', error);

pause  = reading.pause.bind(reading);
resume = reading.resume.bind(reading);

function readLn(line){
  pause();
  row = lineToArray(line);
  return channel.isBlocked
       ? channel.once('unblocked', readRow)
       : readRow();
}

function readRow(){
  message.data = parseRow(row);
  msg = new Buffer(JSON.stringify(message));
  return publishMessage().then(resume, publishMessage);
}

function publishMessage(){
  return channel.sendToQueue('finish.crawl.ly', msg, { persistent: true })
       ? Promise.resolve()
       : waitForChannelDrain();
}

function waitForChannelDrain(){
  return new Promise(function promiseChannelDrain(resolve){
    channel.once('drain', resolve);
  });
}
```

**Comment by jcready** [comment]
Also note that in that example I was also checking if the channel is "blocked" or not, and if it is, I'm waiting for it to become `"unblocked"`. This is different from the `"drain"` event.
see here: [unblocked](http://www.squaremobius.net/amqp.node/doc/channel_api.html#toc_15) vs. [drain](http://www.squaremobius.net/amqp.node/doc/channel_api.html#toc_26)

**Pull request merged**